src/Compatibility/Microsoft.DotNet.PackageValidation/Validators/CompatibleTFMValidator.cs

```diff
@@ -51,67 +48,67 @@ public void Validate(Package package)
                         DiagnosticIds.CompatibleRuntimeRidLessAsset,
                         Resources.NoCompatibleRuntimeAsset,
                         framework.ToString());
                 }
-                else
+                // Invoke ApiCompat to compare the compile time asset with the runtime asset if they are not the same assembly.
+                else if (option.RunApiCompat && compileTimeAsset.Path != runtimeAsset.Path)
                 {
-                    if (_runApiCompat && compileTimeAsset.Path != runtimeAsset.Path)
-                    {
-                        string header = string.Format(Resources.ApiCompatibilityHeader, compileTimeAsset.Path, runtimeAsset.Path);
-                        _apiCompatRunner.QueueApiCompatFromContentItem(package.PackageId, compileTimeAsset, runtimeAsset, header);
-                    }
+                    string header = string.Format(Resources.ApiCompatibilityHeader, compileTimeAsset.Path, runtimeAsset.Path);
+                    apiCompatRunner.QueueApiCompatFromContentItem(option.Package.PackageId, compileTimeAsset, runtimeAsset, header);
                 }
 
-                foreach (string rid in package.Rids.Where(t => IsSupportedRidTargetFrameworkPair(framework, t)))
+                foreach (string rid in option.Package.Rids.Where(t => IsSupportedRidTargetFrameworkPair(framework, t)))
                 {
-                    runtimeAsset = package.FindBestRuntimeAssetForFrameworkAndRuntime(framework, rid);
-                    if (runtimeAsset == null)
+                    ContentItem? runtimeRidSpecificAsset = option.Package.FindBestRuntimeAssetForFrameworkAndRuntime(framework, rid);
+                    if (runtimeRidSpecificAsset == null)
                     {
                         _log.LogError(
-                            new Suppression { DiagnosticId = DiagnosticIds.CompatibleRuntimeRidSpecificAsset, Target = framework.ToString() + "-" + rid },
+                            new Suppression(DiagnosticIds.CompatibleRuntimeRidSpecificAsset) { Target = framework.ToString() + "-" + rid },
                             DiagnosticIds.CompatibleRuntimeRidSpecificAsset,
                             Resources.NoCompatibleRidSpecificRuntimeAsset,
                             framework.ToString(),
                             rid);
                     }
-                    else
+                    // Invoke ApiCompat to compare the compile time asset with the runtime specific asset if they are not the same and
+                    // if the comparison hasn't already happened (when the runtime asset is the same as the runtime specific asset).
+                    else if (option.RunApiCompat &&
+                        compileTimeAsset.Path != runtimeRidSpecificAsset.Path &&
+                        (runtimeAsset == null || runtimeAsset.Path != runtimeRidSpecificAsset.Path))
```

Comment:
@joperezr this is the only change in the PR that removed an apicompat runner work item because it was already queued (in L75) to avoid unnecessary work. I admit that this change should have been made in a separate PR and I will keep that in mind for future changes. That said, I don't think that this change should be reverted now and be brought back in a separate PR. If we would trace back a regression because of this change we can simply remove this line.

Reply:
Ok, as long as you are sure this is the only change I guess that is fine. My main point here is: what happens if there is a regression with this line that we discover in the future, we need to fix it, and you are not around? Tracking which line to revert would be a lot harder. Keeping functional changes in separate PRs with a lengthly description of the changes and dedicated review helps with this, since if a regression is found in the future and you (or whoever authors the change) are not around, then finding the culprit and knowing what to revert will be much easier.

Anyway, I'm fine with this instance being kept, but it would be good if for future changes we try to have several atomic and smaller PRs that are more specific to some change, than globing many changes on one single one 😉